libraries/lib-cloud-audiocom/OAuthService.cpp

```diff
@@ -0,0 +1,370 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  UploadService.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include "OAuthService.h"
+
+#include <cassert>
+#include <cctype>
+
+#include <rapidjson/document.h>
+#include <rapidjson/writer.h>
+
+#include "CodeConversions.h"
+#include "Prefs.h"
+
+#include "IResponse.h"
+#include "NetworkManager.h"
+#include "Request.h"
+
+#include "ServiceConfig.h"
+
+#include "UrlDecode.h"
+
+#include "BasicUI.h"
+
+namespace cloud::audiocom
+{
+namespace
+{
+
+StringSetting refreshToken { L"/cloud/audiocom/refreshToken", "" };
+
+const std::string_view uriPrefix = "audacity://link";
+const std::string_view usernamePrefix = "username=";
+const std::string_view passwordPrefix = "password=";
+const std::string_view tokenPrefix = "token=";
+const std::string_view authorizationCodePrefix = "authorization_code=";
+
+void WriteCommonFields(
+   rapidjson::Document& document, std::string_view grantType, std::string_view scope)
+{
+   using namespace rapidjson;
+
+   document.AddMember(
+      "grant_type", StringRef(grantType.data(), grantType.size()),
+      document.GetAllocator());
+
+   const auto clientID = GetServiceConfig().GetOAuthClientID();
+
+   document.AddMember(
+      "client_id", StringRef(clientID.data(), clientID.size()),
+      document.GetAllocator());
+
+   document.AddMember(
+      "client_secret", StringRef("shKqnY2sLTfRK7hztwzNEVxnmhJfOy1i"),
+      document.GetAllocator());
+
+   document.AddMember(
+      "scope", StringRef(scope.data(), scope.size()), document.GetAllocator());
+}
+
+bool IsPrefixed(std::string_view hay, std::string_view prefix)
+{
+   if (hay.length() < prefix.length())
+      return false;
+
+   return std::equal(
+      hay.begin(), hay.begin() + prefix.length(), prefix.begin(), prefix.end(),
+      [](auto a, auto b) { return std::tolower(a) == b; });
```

Comment:
Perhaps overload 3 might express it more clearly: https://en.cppreference.com/w/cpp/algorithm/mismatch